File: two_pointers/three_sum.rb
# frozen_string_literal: true

# Time complexity: O(n^2)
# Space complexity: O(1) or O(n)
def three_sum(nums)
  res = []
  sorted = nums.sort
  length = nums.length

  (0..length - 3).each do |i1|
    return res.keys if sorted[i1].positive?
    next if i1.positive? && sorted[i1] == sorted[i1 - 1]

    i2 = i1 + 1
    i3 = length - 1

    while i2 < i3
      triplet = [sorted[i1], sorted[i2], sorted[i3]]
      sum = triplet.reduce(&:+)

      if sum.positive?
        i3 -= 1
      elsif sum.negative?
        i2 += 1
      else
        res << triplet
        i2 += 1
        i2 += 1 while sorted[i2] == sorted[i2 - 1] && i2 < i3
      end
    end
  end

  res
end

File: two_pointers/three_sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        length = len(nums)
        nums.sort()

        for i, ni in enumerate(nums):
            if i > 0 and nums[i - 1] == ni:
                continue

            l, r = i + 1, length - 1
            while l < r:
                total = nums[i] + nums[l] + nums[r]

                if total < 0:
                    l += 1
                elif total > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    while (nums[l - 1] == nums[l] and l < r):
                        l += 1

        return res
    
    def threeSumNaive(self, nums: List[int]) -> List[List[int]]:
        res = []
        incl = set()
        length = len(nums)

        for i in range(length - 2):
            for j in range(i + 1, length - 1):
                for k in range(j + 1, length):
                    if (nums[i] + nums[j] + nums[k]) == 0:
                        tmp = tuple(sorted([nums[i], nums[j], nums[k]]))
                        if tmp not in incl:
                            res.append([nums[i], nums[j], nums[k]])
                            incl.add(tmp)

        return res